src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/TemporaryArray`1.cs

```diff
@@ -0,0 +1,313 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Collections
+{
+    /// <summary>
+    /// Provides temporary storage for a collection of elements. This type is optimized for handling of small
+    /// collections, particularly for cases where the collection will eventually be discarded or used to produce an
+    /// <see cref="ImmutableArray{T}"/>.
+    /// </summary>
+    /// <remarks>
+    /// This type stores small collections on the stack, with the ability to transition to dynamic storage if/when
+    /// larger number of elements are added.
+    /// </remarks>
+    /// <typeparam name="T">The type of elements stored in the collection.</typeparam>
+    [NonCopyable]
+    [StructLayout(LayoutKind.Sequential)]
+    internal struct TemporaryArray<T> : IDisposable
```

Comment:
@sharwell I'm wondering if it is at all possible / preferable to use InlineArray for this?

Reply:
@alrz `InlineArray` wouldn't be a complete replacement for this, but it would certainly be a viable replacement for the `_item0`...`_item3` fields.